File: android/app/src/main/kotlin/com/thecityandthebike/ui/viewmodel/DeleteAccountViewModel.kt
package com.thecityandthebike.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.thecityandthebike.data.model.ApiResult
import com.thecityandthebike.data.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class DeleteAccountState(
    val isDeleting: Boolean = false,
    val isDeleted: Boolean = false,
    val error: String? = null
)

@HiltViewModel
class DeleteAccountViewModel @Inject constructor(
    private val authRepository: AuthRepository
) : ViewModel() {

    private val _state = MutableStateFlow(DeleteAccountState())
    val state: StateFlow<DeleteAccountState> = _state.asStateFlow()

    fun deleteAccount() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isDeleting = true, error = null)
            when (val result = authRepository.deleteAccount()) {
                is ApiResult.Success -> {
                    _state.value = _state.value.copy(isDeleting = false, isDeleted = true)
                }
                is ApiResult.Error -> {
                    _state.value = _state.value.copy(
                        isDeleting = false,
                        error = result.error.displayMessage
                    )
                }
            }
        }
    }

    fun clearError() {
        _state.value = _state.value.copy(error = null)
    }
}

File: ios/Packages/TCATBProfile/Sources/TCATBProfile/DeleteAccountViewModel.swift
import Foundation
import Observation

@Observable
public final class DeleteAccountViewModel {
    public private(set) var isDeleting = false
    public private(set) var isDeleted = false
    public private(set) var error: String?

    private let apiClient: ProfileAPIClient

    public init(apiClient: ProfileAPIClient) {
        self.apiClient = apiClient
    }

    public func deleteAccount() async {
        isDeleting = true
        error = nil

        do {
            try await apiClient.deleteAccount()
            isDeleted = true
        } catch {
            self.error = error.localizedDescription
        }

        isDeleting = false
    }

    public func clearError() {
        error = nil
    }
}
